### ISSUE TITLE
Revamp Ballot Privacy WAS: Revoke ballot is too CPU-intensive

### ISSUE THREAD

**Issue by jberkus** [elekto-io/elekto#49]
## The problem

Currently our "revoke ballot" code uses check_password_hash() to locate a voters prior votes, like this:

```
    voter = F.g.user.username + '+' + F.request.form['password']
    ballots = [b for b in e.ballots if check_password_hash(b.voter, voter)]
```

We do this because we need to check the voter-ballot link in a way that preserves the anonymity of voting.  However, the above amounts to running check_password_hash() once per voter/candidate combination.  As we've discovered in Kubernetes, if you have 90 ballots cast with 13 candidates, that's over 1100 checks taking around 75 seconds, during with a whole CPU is pinned.  

While we could change the code somewhat to make the check run once per voter only, that's still not a solution; elections like Kubernetes have over 700 voters, and check_password_hash() is just too slow.  We'll need another approach.

## The solution

First thought is to look at the Helios code to see if they have a more efficient way of doing this.

Other ideas???

**Comment by BenTheElder** [comment]
Can we do something like:
```
    voter = F.g.user.username + '+' + F.request.form['password']
    voter_hash = generate_password_hash(voter)
    ballots = [b for b in e.ballots if b.voter == voter_hash]
```

I must be missing something, it seems like the voter field on a ballot is already the pre-computed hash, it seems like this should basically just be a cheap string comparison per ballot?

https://github.com/elekto-io/elekto/blob/46597d459be9600e1fe39309392a5186e6e04fd9/elekto/models/sql.py#L173

**Comment by PushkarJ** [comment]
Assuming revoke ballot is needed only when editing vote, we could move the voters who want their votes edited later, into a separate table. This will allow us to save CPU time, by skipping the ones who do not create a passphrase to allow editing later. 

It does not solve the problem completely but atleast improves the performance in some way

**Comment by BenTheElder** [comment]
> generate_password_hash will create a new salt and a new hash everytime it is used. So for same input voter the output of the func will be different everytime. Therefore checking equality of hashes does not work here.

Thanks, this is more apparent for me at least reading the sources which appear to be here:
https://github.com/pallets/werkzeug/blob/d2625f545f53252314aee0c36818f52a52f7d321/src/werkzeug/security.py

So the salt is random per ballot entry into the db, and then we're computing the hash for the voter revoking against every existing ballot's salt ...

> I don't really see the point in workarounds. For the current kubernetes election we're just going to disable ballot revokation. For later elections, we'll want to have an actual fix in place.

+1, I couldn't find docs about the current approach but what about this for next time?

Instead of hashing (handle, password) to identify ballots we can split two tables:
1. `ballot_random_id => ballot_contents`
2. `handle => encrypt(ballot_random_id, password)`

Don't put _any_ reference to the user in the ballots table, and instead use the second handle => password encrypted ballot ID table.

Tabulating still won't have any idea whose ballot is whose since the second table is useless without the passwords but we can compute results from the first table alone. When presented with (handle, password) we can then and only then identify and mutate a specific ballot.

NOTE: In this `ballot_contents` is hand waved to be the full set of ranked choices, not N different ballot entries, it looks like this would need some additional refactoring.

**Comment by jberkus** [comment]
@BenTheElder that reduces the scope of the problem, but I'd really much rather find a solution that allows the ballot lookups to be indexed.  Otherwise there's always a fairly low scalability limit.

**Comment by BenTheElder** [comment]
Er why would we need to do that?

Aren't we given (handle, password)?

We can lookup the exact encrypted ballot ID by the handle in table one and then decrypt it with the password to lookup the exact ballot in table 2.

Hashing is dropped entirely in favor of encrypting the link between handle and ballot.

**Comment by jberkus** [comment]
OH!  Dammit, I had to mock this one out to figure it out.  Let me make sure I got it right. In my example below I'm gonna use much shorter numbers than we would actually.

So I vote.  My ballot gets created with the randomly chosen number 107 and I use the passphrase "kubernetes".

In the ballots table we get:

```
107, vote1
107, vote2
...
```

And in the handles table we get:

```
jberkus, fb9a2abc26...
```

Where "fb9a2abc26.." is the result of encrypting "107" using a key derived from the password "kubernetes" .

So if I want to revoke, I enter my passphrase.  Application looks me up in Handles by index, then tries to decrypt the value using a key derived from the passphrase I enter.  If it fails, I get a "wrong passphrase"; if it succeeds, then it has 107 and can delete my ballot.

Did I follow that correctly?

**Comment by jberkus** [comment]
Fernet+KDF is pretty foolproof.  According to industry literature, you can only get a collision there if you hit on a short passphrase which happens to be exactly the same as a KDF digest of a really long passphrase (over 64 bytes), and it's improbable enough that you really need to know the long passphrase in the first place to make it work (like, 1 in 10^64).

**Comment by jberkus** [comment]
FYI, switching to pyNaCL for encryption due to feedback from actual cryptography folks.  It's more current.

**Comment by jberkus** [comment]
@riking we looked over this, and currently there is no per-vote isolation.  Ballots need to be bundled for the preference voting algorithm to work -- and they are in the old code.  So we're going to revert to the idea of having a single per-ballot (not per vote)  unique ID and encrypting that, unless someone knows a way around it.

**Comment by jberkus** [comment]
So, new schema is:
```
    __tablename__ = 'voter'

    id = S.Column(S.Integer, primary_key=True)
    user_id = S.Column(S.Integer, S.ForeignKey('user.id', ondelete="CASCADE"))
    election_id = S.Column(S.Integer, S.ForeignKey('election.id', ondelete="CASCADE"))
    created_at = S.Column(S.DateTime, default=S.func.now())
    updated_at = S.Column(S.DateTime, default=S.func.now())
    salt = S.Column(S.Integer, nullable=False)
    key = S.Column(S.String(255), nullable=False)
 
    __tablename__ = 'ballot'

    # Attributes
    id = S.Column(S.GUID, primary_key=True)
    election_id = S.Column(S.Integer, S.ForeignKey('election.id', ondelete='CASCADE'))
    rank = S.Column(S.Integer, default=100000000)
    candidate = S.Column(S.String(255), nullable=False)
    voter = S.Column(S.GUID, nullable=False)
```

ballot.voter now contains a GUID that is generated on a per-ballot (groupable for the multiple votes in a ballot), the same as the value in ballot.voter is now.  voter.key contains the encrypted version of that GUID, and voter.salt has the randomly generated salt for it.  The advantage of these GUIDs is that they make reconstruction of the ballots via row ordering very hard.

We should also strip off created_on and updated_on from ballot, given that they make row ordering and timing attacks easier.

**Comment by jberkus** [comment]
BTW, per discussion here are the ballot privacy vectors we're worried about:

1. Election officers eavesdropping on votes
2. Sysadmins eavesdropping on votes
3. Stolen database backups

**Comment by jberkus** [comment]
Resolved in 0.6.0
